File: BOJ/BOJ_1388_바닥장식.py
import sys
sys.setrecursionlimit(10**6)

N, M = map(int, input().split())
graph = [input() for _ in range(N)]
visited = [[0] * M for _ in range(N)]
dy = [-1, 1, 0, 0]
dx = [0, 0, -1, 1]
cnt = 0

def dfs(y, x, c):
    visited[y][x] = 1
    if c == '|':
        for i in range(2):
            ny, nx = y + dy[i], x + dx[i]
            if 0 <= ny < N and 0 <= nx < M and graph[ny][nx] == c and not(visited[ny][nx]):
                dfs(ny, nx, c)
    elif c == '-':
        for i in range(2, 4):
            ny, nx = y + dy[i], x + dx[i]
            if 0 <= ny < N and 0 <= nx < M and graph[ny][nx] == c and not(visited[ny][nx]):
                dfs(ny, nx, c)

for i in range(N):
    for j in range(M):
        if not(visited[i][j]):
            cnt += 1
            dfs(i, j, graph[i][j])

print(cnt)

File: BOJ/BOJ_1388_바닥장식.cpp
#include <bits/stdc++.h>
using namespace std;

int N, M, cnt;
char graph[101][101];
bool visited[101][101];
int dy[4] = {-1, 1, 0, 0};
int dx[4] = {0, 0, -1, 1};

void init() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);
}

void DFS(int y, int x, char c) {
    visited[y][x] = true;
    if (c == '|') {
        for (int i = 0; i < 2; i++) {
            int ny = y + dy[i];
            int nx = x + dx[i];
            if (0 <= ny && ny < N && 0 <= nx && nx < M && graph[ny][nx] == c && !(visited[ny][nx])) {
                DFS(ny, nx, c);
            }
        }
    }
    else if (c == '-') {
        for (int i = 2; i < 4; i++) {
            int ny = y + dy[i];
            int nx = x + dx[i];
            if (0 <= ny && ny < N && 0 <= nx && nx < M && graph[ny][nx] == c && !(visited[ny][nx])) {
                DFS(ny, nx, c);
            }
        }
    }
}

int main(int argc, char const *argv[]) {
    init();
    cin >> N >> M;
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < M; j++) {
            cin >> graph[i][j];
        }
    }
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < M; j++) {
            if (!(visited[i][j])) {
                cnt++;
                DFS(i, j, graph[i][j]);
            }
        }
    }
    cout << cnt << '\n';
    return 0;
}